framework/src/eventloop/eventloop_event.c

```diff
@@ -81,283 +79,321 @@ static event_group_t *get_event_group(int type)
-		eldbg("Invalid Parameter\n");
-		return EVENTLOOP_INVALID_PARAM;
-	}
 
-	event_group = get_event_group(type);
-	if (event_group != NULL) {
-		ptr = (event_node_t *)sq_peek(&event_group->event_list);
-		while (ptr != NULL && ptr->handle != NULL) {
-			event_data_t *cb_data = (event_data_t *)ptr->handle->data;
-			cb_data->event_in = type;
-			if (data_size == 0) {
-				cb_data->event_data = NULL;
-			} else {
-				cb_data->event_data = EL_ALLOC(data_size);
-				if (cb_data->event_data == NULL) {
-					eldbg("Failed to allocate callback info\n");
-					return EVENTLOOP_OUT_OF_MEMORY;
+	for (i = 1; i < EL_EVENT_MAX; i++) {
+		event_group = get_event_group(i);
+		if (event_group != NULL) {
+			handle = (el_handle_t *)sq_peek(&event_group->event_list);
+			while (handle) {
+				if (handle->pid == getpid() && handle->mode == EL_MODE_DEFAULT) {
+					sq_rem((FAR sq_entry_t *)handle, &event_group->event_list);
+					eventloop_decrease_app_handle_cnt(handle->pid);
+					eventloop_clean_event_handle(handle);
+					break;
 				}
-				memcpy(cb_data->event_data, event_data, data_size);
+				handle = (el_handle_t *)sq_next(handle);
 			}
-			/* Send signal to task which registered event */
-			int ret = kill(cb_data->pid, SIGEL_EVENT);
-			if (ret < 0) {
-				eldbg("kill failed %d \n", errno);
-				if (cb_data->event_data != NULL) {
-					EL_FREE(cb_data->event_data);
-					cb_data->event_data = NULL;
-				}
+			if (sq_empty(&event_group->event_list)) {
+				sq_rem((FAR sq_entry_t *)event_group, &g_event_list);
+				EL_FREE(event_group);
 			}
-			ptr = (event_node_t *)sq_next(ptr);
 		}
 	}
+}
 
-	return OK;
+/* Event callback function called in user side */
+void eventloop_event_user_callback(el_handle_t *handle)
+{
+	int ret;
+	el_event_t *event;
+
+	if (handle == NULL) {
+		return;
+	}
+
+	elvdbg("[%d] Event Callback!\n", getpid());
+
+	event = (el_event_t *)&handle->data.event;
+	if (event->cb_func) {
+		/* Execute timer callback */
+		ret = event->cb_func(event->cb_data, event->event_data);
+		if (ret == EVENTLOOP_CALLBACK_STOP) {
+			eventloop_unregister_event_type(event->event_type);
+		}
+	}
+
+	eventloop_clean_event_handle(handle);
 }
 
 el_event_t *eventloop_add_event_handler(int type, event_callback func, void *data)
 {
 	int ret;
-	el_loop_t *loop;
-	el_event_t *handle;
-	event_data_t *event_cb;
+	el_handle_t *handle;
+	el_event_t *event_handler;
 
 	if (type < 0 || type >= EL_EVENT_MAX || func == NULL) {
 		eldbg("Invalid Parameter\n");
 		return NULL;
 	}
 
-	loop = get_app_loop();
-	if (loop == NULL) {
-		eldbg("Failed to get loop\n");
+	handle = (el_handle_t *)EL_ALLOC(sizeof(el_handle_t));
+	if (handle == NULL) {
 		return NULL;
 	}
-
-	handle = (el_event_t *)EL_ALLOC(sizeof(el_event_t));
-	if (handle == NULL) {
-		eldbg("Failed to allocate event\n");
+	handle->pid = getpid();
+	handle->type = EL_HANDLE_TYPE_EVENT;
+	handle->mode = EL_MODE_DEFAULT;
+
+	event_handler = (el_event_t *)&handle->data.event;

```

Comment:
I think you'd better check that event_handler is NULL or not

Reply:
I don't think so because it is not dynamically allocated.